server/api/totp/activate/controller.test.js

```diff
@@ -0,0 +1,299 @@
+        .set('Authorization', `Bearer ${wileSession.accessToken}`)
+        .send({
+          userId: wileUser.id,
+          secret: 'N34CXTAEDWWIETTGN7P7HGFVM2CPGAG2',
+        });
+
+      expect(res.status).toBe(200);
+      expect(res.body).toMatchObject({
+        ok: false,
+        error: {
+          code: 400,
+          message: 'Invalid request body',
+          details: expect.any(Array),
+        },
+      });
+      expect(res.body.error.details[0].path).toEqual(['token']);
+      expect(res.body.error.details[0].type).toBe('any.required');
+    });
+
+    test('returns error when `token` body param is invalid', async () => {
+      const res = await request(server)
+        .post('/api/totp.activate')
+        .set('Authorization', `Bearer ${wileSession.accessToken}`)
+        .send({
+          userId: wileUser.id,
+          secret: 'N34CXTAEDWWIETTGN7P7HGFVM2CPGAG2',
+          token: '1234567', // i.e. more chars than expected
+        });
+
+      expect(res.status).toBe(200);
+      expect(res.body).toMatchObject({
+        ok: false,
+        error: {
+          code: 400,
+          message: 'Invalid request body',
+          details: expect.any(Array),
+        },
+      });
+      expect(res.body.error.details[0].path).toEqual(['token']);
+      expect(res.body.error.details[0].type).toBe('string.length');
+    });
+
+    test('returns error when secret key does not exist in token collection', async () => {
+      const res = await request(server)
+        .post('/api/totp.activate')
+        .set('Authorization', `Bearer ${wileSession.accessToken}`)
+        .send({
+          userId: wileUser.id,
+          secret: 'N34CXTAEDWWIETTGN7P7HGFVM2CPGAG2',
+          token: '123456',
+        });
+
+      expect(res.status).toBe(200);
+      expect(res.body).toMatchObject({
+        ok: false,
+        error: {
+          code: 10027,
+          message: 'Secret key does not exist or has already expired',
+        },
+      });
+    });
+
+    test('returns error when secret key exists but is associated with another user', async () => {
+      const TokenModel = ctx.db.model('Token');
+      const TOTPModel = ctx.db.model('TOTP');
+
+      const secret = TOTPModel.generateSecret();
+      const tokenRecord = await TokenModel.create({
+        secret,
+        type: 'TOTP_ENROLL',
+        user: ObjectId('507f191e810c19729de860ea'), // i.e. some random object ID
+        org: null,
+        expiresAt: addSeconds(new Date(), 1000),
+      });
+
+      const res = await request(server)
+        .post('/api/totp.activate')
+        .set('Authorization', `Bearer ${wileSession.accessToken}`)
+        .send({
+          userId: wileUser.id,
+          secret,
+          token: '123456',
+        });
+
+      expect(res.status).toBe(200);
+      expect(res.body).toMatchObject({
+        ok: false,
+        error: {
+          code: 10027,
+          message: 'Secret key does not exist or has already expired',
```

Comment:
This message is confusing me a little. If i try to activate a secret key, i get back that this secret does not exist?

Reply:
Fixed.